### PR TITLE
zkasm: i32 overflows

### PR DESCRIPTION
**Pull request by MCJOHN974**

### CODE REVIEW

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 1]
```diff
@@ -455,29 +456,50 @@ impl MachInstEmit for Inst {
...
+                // no register to store rs2 =(
+                put_string("B :MSTORE(SP)\n", sink);
+                // E / 2 ** 32 => A
+                put_string("0 => D\n", sink);
+                put_string("0 => C\n", sink);
+                put_string("4294967296n => B\n", sink);
+                put_string("$${var _divArith = E / B}\n", sink);
+                // Intentionnaly ignore rem here because if rem != 0 mean something goes wrong
+                put_string("${_divArith} => A\n", sink);
+                put_string("E:ARITH\n", sink);
```
Comment (nagisa) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:481]: I’ve been wondering for a while now, but why is it necessary to use these `var` and all of the intermediate steps, when this should basically be expressible as the following snippet?

```
E / 4294967296n => A
E: ARITH
```

Comment (MCJOHN974) [reply cranelift/codegen/src/isa/zkasm/inst/emit.rs]: Hmm, yes, `` ${E / B} => A`` allowed. But `` ${E / 4294967296n} => A`` not allowed, so we can't free ``B`` from clobbering. Removed unnecessary ``var`` [here](https://github.com/near/wasmtime/pull/79/commits/9cabc52382570fd65474f0dd3577639b5b2c80ff)

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 2]
```diff
@@ -503,27 +503,103 @@ impl MachInstEmit for Inst {
...
-                // Same as in MulArith about D register
                 put_string("0 => D\n", sink);
                 put_string("$${var _divArith = E / B}\n", sink);
                 put_string("$${var _remArith = E % B}\n", sink);
                 put_string("${_divArith} => A\n", sink);
                 put_string("${_remArith} => C\n", sink);
                 put_string("E:ARITH\n", sink);
             }
+            // Rem32 is quite difficult opcode. Here we need calculate
+            // ((op1 / 2**32) % (op2 / 2**32)) * 2**32
```
Comment (MCJOHN974) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:536]: Also, there are many different ways to implement this opcode, using different register as inputs and outputs for different operations. And I believe this one is not the best. But I think anyway this operation is quite difficult and optimization will be relatively small, measuring in lines of zkasm code. So I suggest keep this version for now as working version, and in future think about more optimal implementation, if it will be necessary

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs (hunk 1 above)
Comment (nagisa) [reply cranelift/codegen/src/isa/zkasm/inst/emit.rs]: It does work without the `n` within `{}` though.

```
  4294967296n => C
  ${C / 4294967296} => A
  1n => B
  B :ASSERT
```

Executes just fine. Still cannot avoid clobbering B regardless as it is used as an input by the `ARITH` that follows. But my question was largely about the unnecessary complexity in the generated code, and it seems like it might not be necessary.

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 3]
```diff
@@ -1366,10 +1368,24 @@ impl MachInstEmit for Inst {
...
+                    put_string(&format!("{one} => B\n"), sink);
                     put_string(&format!("$ => {} :XOR\n", reg_name(rd.to_reg())), sink);
                 }
 
+                // Result of comparing operations in wasm for both i32 and i64
+                // is i32. So, we need to multiply it by 2**32
+                put_string("4294967296n => B\n", sink);
+                put_string("0 => D\n", sink);
+                put_string("0 => C\n", sink);
+                put_string("${A * B} => A :ARITH\n", sink);
```
Comment (nagisa) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:1387]: My intuition suggests using jumps here would be faster if only by the virtue of forcing fewer register spills.

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 4]
```diff
@@ -1366,10 +1368,24 @@ impl MachInstEmit for Inst {
                 put_string(&format!("$ => {} :{opcode}\n", reg_name(rd.to_reg())), sink);
 
                 if cc == IntCC::NotEqual {
-                    put_string("1 => B\n", sink);
+                    let one: u64 = if ty == types::I32 {
+                        1
+                    } else if ty == types::I64 {
+                        4294967296
```
Comment (nagisa) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:1374]: Not sure I follow why this needs to be done. If comparison puts a `1` in `A`, then after `XOR` the value in `A` still stays non-zero (i.e. truthy.)

File: cranelift/codegen/src/isa/zkasm/inst/mod.rs [hunk 5]
```diff
@@ -307,6 +316,17 @@ fn zkasm_get_operands<F: Fn(VReg) -> VReg>(inst: &Inst, collector: &mut OperandC
             collector.reg_clobbers(clobbered);
             collector.reg_fixed_def(rd, c0());
         }
+        &Inst::RemArith32 { rd, rs1, rs2, .. } => {
```
Comment (nagisa) [inline cranelift/codegen/src/isa/zkasm/inst/mod.rs:319]: As this and the branch above do not differ, merge them:

```rust
Inst::RemArith { rd, rs1, rs2, .. } | Inst::RemArith32 { rd, rs1, rs2, .. } => { ... }
```

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs (hunk 3 above)
Comment (nagisa) [reply cranelift/codegen/src/isa/zkasm/inst/emit.rs:1387]: (this is non-blocking for landing)

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs (hunk 4 above)
Comment (MCJOHN974) [reply cranelift/codegen/src/isa/zkasm/inst/emit.rs:1374]: afaik it legal to write something like this:
```
i32.const 3
i32.const 5
i32.neq
i32.const 1
i32.add
i32.const 2
call $assert_eq
```

And in general comp operations return always 1 or 0 as i32 and it will be nice to keep this numbers corresponding our i32 representation

Comment (nagisa) [reply cranelift/codegen/src/isa/zkasm/inst/emit.rs:1374]: You are correct of course, but it will still produce the same `1` or `0` in register `A` regardless of whether 32-bit integers or 64-bit integers are being compared. So I’m still not following what this `2^32` is achieving here.

File: cranelift/codegen/src/isa/zkasm/inst/mod.rs (hunk 5 above)
Comment (nagisa) [reply cranelift/codegen/src/isa/zkasm/inst/mod.rs:319]: Hm, so forward looking I was thinking that we’d perhaps do something to abstract away some common parts about these different instructions to reduce duplication, but I guess you’re right that it makes no sense to do so especially if we’re looking at these code paths diverging even further as development proceeds.

**Comment by MCJOHN974** [comment]
Hm, I don't understand why jobs don't run, @akashin, maybe we changed something that causes it?

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -745,7 +745,7 @@ jobs:
     name: Record the result of testing and building steps
     runs-on: ubuntu-latest
     needs:
-      - test_zkasm
+      - test_zkasm_64
```
Comment (nagisa) [inline .github/workflows/main.yml:748]: Is the cause for CI failures :)

Comment (MCJOHN974) [reply .github/workflows/main.yml]: Hmm, strange, while merging I was accepting incoming changes everywhere, how this wasn't changed due to mege?.. But it works, thanks!

**Comment by MCJOHN974** [comment]
```
18446744073709551616 (0x10000000000000000) != 115792089237316195423570985008687907853269984665640564039457584007913129639936 (0x10000000000000000000000000000000000000000000000000000000000000000)
```
Looks like we returned to 256-bit prover (=

**Comment by MCJOHN974** [comment]
I checked changes this PR made and believe this kind of error also related to new prover?.. @akashin 
```Error: Program terminated with registers A, D, E, PC, HASHPOS, RR, RCX, zkPC not set to zero```

**Comment by akashin** [comment]
> I checked changes this PR made and believe this kind of error also related to new prover?.. @akashin `Error: Program terminated with registers A, D, E, PC, HASHPOS, RR, RCX, zkPC not set to zero`



> I checked changes this PR made and believe this kind of error also related to new prover?.. @akashin `Error: Program terminated with registers A, D, E, PC, HASHPOS, RR, RCX, zkPC not set to zero`

The failing tests include `fibonacci.zkasm`, so I imagine this should fail in the main branch as well. Let me try to reproduce this.

**Comment by akashin** [comment]
Ok, the tests pass on the main branch, so the issue must be a combination of change to the testing infra and changes introduced in the PR.
I suggest we investigate why exactly the tests fail:
- Check the number of counters in failing tests by running `run-tests-zkasm.js` with `--verbose` flag
- Inspect the generated `.zkasm` files to find what is different and might cause this issue
- Compare the [PIL configs](https://github.com/near/wasmtime/blob/main/tests/zkasm/run-tests-zkasm.js#L58) used before and now - I can do this

Our goal should also be instrumenting the code in a way that makes debugging this failure easier.

**Comment by MCJOHN974** [comment]
> Check the number of counters in failing tests by running run-tests-zkasm.js with --verbose flag

For failing files it don't show counters, for not failing it show something like this:
```
   --> pass /home/mcjohn974/pagoda/wasmtime/cranelift/zkasm_data/generated/div.zkasm
   --> verbose
        --> counters
{
  cntArith: 0n,
  cntBinary: 0n,
  cntKeccakF: 0n,
  cntMemAlign: 0n,
  cntSteps: 29
}
        --> outputs
undefined
        --> logs
{}
```
> Inspect the generated .zkasm files to find what is different and might cause this issue

I did it and I believe it's ok. And in previous testing system it worked.

**Comment by MCJOHN974** [comment]
@akashin can you please help with this? I think it is related to new prover

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 7]
```diff
@@ -479,52 +512,104 @@ impl MachInstEmit for Inst {
...
-                //);
-                // put_string(
-                //   &format!("${{_mulArith % (2 ** 64)}} => {} :ARITH\n", reg_name(rd.to_reg())),
-                //    sink,
-                //);
-                // For now we will just clear D in case it was something in it
                 put_string("0 => D\n", sink);
                 put_string("0 => C\n", sink);
                 put_string("$${var _mulArith = A * B}\n", sink);
+                put_string("${_mulArith / 18446744073709551616} => D\n", sink);
```
Comment (akashin) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:518]: Let's use bit shifts here to make this code easier to understand. I.e.
```
${_mulArith >> 64} => D
```

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 8]
```diff
@@ -479,52 +512,104 @@ impl MachInstEmit for Inst {
...
                 put_string(
-                    &format!("${{_mulArith}} => {} :ARITH\n", reg_name(rd.to_reg())),
+                    &format!(
+                        "${{_mulArith % 18446744073709551616}} => {} :ARITH\n",
```
Comment (akashin) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:521]: Looking at ZK EVM ROM, I see that we don't use modulo there https://github.com/0xPolygonHermez/zkevm-rom/blob/c17c1664a654e27de67bcbcbe6b6ae44c5b93274/main/utils.zkasm#L594

I suspect this is because it happens automatically when trying to put the value in the register. Might be worth asking Polygon team about this.

In any case, let's use bitwise & here instead of a large constant to make this code easier to read.

File: cranelift/codegen/src/isa/zkasm/lower.isle [hunk 9]
```diff
@@ -31,8 +31,8 @@
 (rule 0 (lower (has_type (ty_int_ref_scalar_64 ty) (iadd x y)))
   (rv_add x y))
 
-(rule 1 (lower (iadd (imm32_from_value x) (imm32_from_value y)))
-  (zk_add x y))
+;; (rule 1 (lower (iadd (imm32_from_value x) (imm32_from_value y)))
+;;   (zk_add x y))
```
Comment (MCJOHN974) [inline cranelift/codegen/src/isa/zkasm/lower.isle:35]: Idk why but this rule was matching
```
i64.const 1
i64.const 1
i64.add
```
Disabled for now, it also increased amount of passing spectests.

File: cranelift/filetests/src/test_zkasm.rs [hunk 10]
```diff
@@ -80,30 +80,25 @@ mod tests {
     /// inside given function.
     /// Label name must begin from label_.
     fn optimize_labels(code: &[&str], func_index: usize) -> Vec<String> {
-        let mut label_definition: HashMap<usize, usize> = HashMap::new();
-        let mut label_uses: HashMap<usize, Vec<usize>> = HashMap::new();
+        let mut label_definition: HashMap<String, usize> = HashMap::new();
+        let mut label_uses: HashMap<String, Vec<usize>> = HashMap::new();
```
Comment (MCJOHN974) [inline cranelift/filetests/src/test_zkasm.rs:84]: Now labels can have same number but different name, like ``label_5`` and ``label_jmpnz_5``, so now we need use whole label name here, not only index

File: ci/zkasm-result.py [hunk 11]
```diff
@@ -46,7 +46,10 @@ def assert_with_csv(status_map):
                 continue
             csv_dict[row[0]] = row[1]
         if csv_dict != status_map:
-            print(f"dict diff = {csv_dict ^ status_map}")
+            diff = set(csv_dict.items()) ^ set(status_map.items())
+            diff_keys = set(map(lambda x: x[0], diff))
+            for key in diff_keys:
+                print(f"Update for test {key}: {csv_dict[key]} => {status_map[key]}")
```
Comment (MCJOHN974) [inline ci/zkasm-result.py:52]: Small diff print enrichment

File: cranelift/codegen/src/isa/zkasm/lower.isle (hunk 9 above)
Comment (nagisa) [reply cranelift/codegen/src/isa/zkasm/lower.isle:35]: Why shouldn’t it have matched? `1` fits into a `32-bit` immediate (which qualifies this for `1 + 1 => R` form, and the result is going to be `2` either way. Hmm… though I guess that will be problematic if this matches for `0xFFFF_FFFF + 2` which… given the unclear limb computation rules does not work. This should be matching for 31-bit immediates…

**Comment by MCJOHN974** [comment]
@akashin @nagisa, Finally fixed ``JMPNZ``! Think it's ready to merge now

File: cranelift/codegen/src/isa/zkasm/inst/emit.rs [hunk 12]
```diff
@@ -957,16 +957,15 @@ impl MachInstEmit for Inst {
...
-                        // let code = kind.emit();
-                        // let code_inverse = kind.inverse().emit().to_le_bytes();
-                        // sink.use_label_at_offset(start_off, label, LabelUse::B12);
-                        // sink.add_cond_branch(start_off, start_off + 4, label, &code_inverse);
-                        // sink.put4(code);
+                        // WARNING: next two put_string MUST be in that order,
+                        // otherwise, if kind.rs1 == A it will lead to infinite loops.
+                        // Idk how to make kind.rs1 always not equal A.
+                        // At least, both A and B marked as clobbered in
+                        // mod.rs, so it will work.
```
Comment (MCJOHN974) [inline cranelift/codegen/src/isa/zkasm/inst/emit.rs:964]: This is not obvious, and I'd like to fix it, but think it will make sense to do in another PR, maybe together with resolving TODO above

File: cranelift/codegen/src/isa/zkasm/lower.isle (hunk 9 above)
Comment (akashin) [reply cranelift/codegen/src/isa/zkasm/lower.isle:35]: It's a good catch. We should aim to restore this logic for 31-bit immediate as Simonas points out, as this is an important optimization - let's file an issue for this and add a TODO here.

File: cranelift/filetests/src/test_zkasm.rs (hunk 10 above)
Comment (akashin) [reply cranelift/filetests/src/test_zkasm.rs:84]: Would it be possible to split this change into a separate PR, given that it is not related to i32 overflows.
This would:
- Allow this change to submitted faster by not being blocked on review of the rest of the code
- Make both changes smaller and easier to review

File: cranelift/codegen/src/isa/zkasm/lower.isle [hunk 13]
```diff
@@ -77,7 +77,7 @@
   (zk_mul x y))
 
 (rule 1 (lower (has_type (fits_in_32 (ty_int ty)) (imul x y)))
-  (zk_mul x y))
+  (zk_mul_32 x y))
```
Comment (akashin) [inline cranelift/codegen/src/isa/zkasm/lower.isle:80]: How about we try to use the super-power of ISLE here and instead of introducing another intermediate rule, directly lower this to `zk_mul` plus `and`:

```
(rule 1 (lower (has_type (fits_in_32 (ty_int ty)) (imul x y)))
  (gen_andi (zk_mul x y) 0xFFFFFFFF))
```

**Review by akashin — CHANGES_REQUESTED** [review]
See my comment on lowering imul without introducing another instruction

File: cranelift/codegen/src/isa/zkasm/lower.isle (hunk 13 above)
Comment (MCJOHN974) [reply cranelift/codegen/src/isa/zkasm/lower.isle:80]: I think it would be better make ISLE improvements in another PR, since this one is quite big already and such changes aren't necessary for first working version of 32-bit computations

File: cranelift/filetests/src/test_zkasm.rs (hunk 10 above)
Comment (MCJOHN974) [reply cranelift/filetests/src/test_zkasm.rs]: Reverted

File: cranelift/codegen/src/isa/zkasm/lower.isle (hunk 13 above)
Comment (akashin) [reply cranelift/codegen/src/isa/zkasm/lower.isle:80]: Sounds good - let's file an issue to investigate pros and cons between these two approaches and work on this in the future PRs.

**Review by akashin — APPROVED** [review]
Good stuff! Let's add TODOs/file issues for any outstanding work here and get this shipped!
